### DISCUSSION TITLE
Examples for listening to "call service" don't work

### DISCUSSION THREAD

**Discussion by brettonw** [custom-components/pyscript#295] [Q&A] [answered]
First off, let me say pyscript has changed my life for the better - I can finally organize automations mostly the way I want and do complex logic without building out huge yaml templates. Thank you to everybody who made this possible.

The problem I'm having is I want to listen for a timer.finished service call from a specific entity, but the event trigger examples with a filter don't work and I can't figure out how to I should do it. The code given in the docs is:

        from homeassistant.const import EVENT_CALL_SERVICE
        
        @event_trigger(EVENT_CALL_SERVICE, "domain == 'light' and service == 'turn_on'")
        def monitor_light_turn_on_service(service_data=None):
            log.info(f"lights.turn_on service called with service_data={service_data}")

and it throws this error:

        hass     | 2022-01-05 10:08:15 ERROR (MainThread) [custom_components.pyscript.function] run_coro: got exception Traceback (most recent call last):
        hass     |   File "/config/custom_components/pyscript/eval.py", line 708, in call
        hass     |     raise TypeError(f"{self.name}() called with unexpected keyword arguments: {unexpected}")
        hass     | TypeError: monitor_light_turn_on_service() called with unexpected keyword arguments: domain, service

I tried making sure it was "light" and not "lights", but that didn't change anything.

Right now, it seems the only way I can filter call service events is to get the service_data and look for my timer entity. Is there a better way?

**Comment by ryaowe** [comment]
Because of the ```domain == 'light' and service == 'turn_on'``` part, pyscript wants to pass ```domain``` and ```service``` to your function. So you should just need to add those to your function definition. Something like:

```
from homeassistant.const import EVENT_CALL_SERVICE
    
@event_trigger(EVENT_CALL_SERVICE, "domain == 'light' and service == 'turn_on'")
def monitor_light_turn_on_service(service_data=None, domain=None, service=None):
    log.info(f"lights.turn_on service called with service_data={service_data}")
```

**Reply by brettonw** [reply]
so I have to add domain?

**Reply by ryaowe** [reply]
If you are including domain in the ```str_expr``` parameter, yeah:

```
An optional str_expr can be used to match the event data, and the trigger will only occur if that expression evaluates to True or non-zero. 

...

When the @event_trigger occurs, those same variables are passed as keyword arguments to the function in case it needs them
```

**Reply by brettonw** [reply]
Indeed, that fixes the function call. Thank you.

However - I'm still stuck, how do I get called when the timer finishes? It doesn't go through call service when it just finishes. Timer "Events" appear to be different things, are they even covered by pyscript?

**Comment by brettonw** [comment]
I've got this to work:

        @event_trigger(EVENT_CALL_SERVICE, "domain == 'timer'")
        def timer_listener(**kwargs):
          service_data = kwargs["service_data"]
          service = kwargs["service"]
          entity_id = service_data["entity_id"][0] if (service_data["entity_id"] is list) else service_data["entity_id"]
          if (entity_id == "timer.my_timer_entity_id"):
              log.info (f"service: {service}")

...But putting the service and service_data in the function arguments causes the above exception.

**Reply by brettonw** [reply]
based on above comments re: adding domain, service, and service_data to the function args, this works:

        @event_trigger(EVENT_CALL_SERVICE, "domain == 'timer'")
        def timer_listener(domain = None, service = None, service_data = None):
          entity_id = service_data["entity_id"][0] if (service_data["entity_id"] is list) else service_data["entity_id"]
          if (entity_id == "timer. my_timer_entity_id"):
              log.info (f"service: {service}")

**Reply by ryaowe** [reply]
You probably want something like:

```
@event_trigger("timer.finished", 'entity_id == "timer.my_timer_entity_id"')
def office_light_timer_service(service_data=None, entity_id=None):
    log.info("timer finished!")
```

**Comment by brettonw** [comment, marked as answer]
Thanks to ryaowe's help, the working result is:

        @event_trigger("timer.finished", "entity_id == 'timer. my_timer_entity_id'")
        def timer_finished(domain=None, entity_id=None):
            log.info(f"timer finished!")
